raft topology: move the topology coordinator to a dedicated file

piodul opened the pull request:
The `topology_coordinator` is a large class (>1000 loc) which resides in an even larger source file (storage_service.cc, ~7800 loc). This PR moves the topology_coordinator class out of the storage_service.cc file in order to improve modularity and recompilation times during development.

As a first step, the `topology_mutation_builder` and `topology_node_mutation_builder` classes are also moved from storage_service.cc to their own, new header/source files as they are an important abstraction used both by the topology coordinator code and some other code in storage_service.cc that won't be moved.

Then, the `topology_coordinator` is moved out. The `topology_coordinator` class is completely hidden in the new topology_coordinator.cc file and can only be started and waited on to finish via the new `run_topology_coordinator` function.

Fixes: scylladb/scylladb#16605

avikivity commented on configure.py:
```diff
@@ -1189,6 +1189,7 @@ def find_ninja():
                 'rust/wasmtime_bindings/src/lib.rs',
                 'utils/to_string.cc',
                 'service/topology_state_machine.cc',
+                'service/topology_mutation.cc',
```
CMakeLists.txt

avikivity commented on configure.py:
```diff
@@ -1190,6 +1190,7 @@ def find_ninja():
                 'utils/to_string.cc',
                 'service/topology_state_machine.cc',
                 'service/topology_mutation.cc',
+                'service/topology_coordinator.cc',
```
This too

kbr-scylla commented on service/topology_coordinator.hh:
```diff
@@ -0,0 +1,69 @@
+
+class topology_coordinator {
+public:
+    virtual ~topology_coordinator() {}
```
Why make it a virtual interface? It suggests that there might be multiple implementations. But there is and always will be only one.

WDYT about using pimpl pattern instead? Like with `group0_guard`.
```cpp
class group0_guard {
    struct impl;
    std::unique_ptr<impl> _impl;

    group0_guard(std::unique_ptr<impl>);
public:
    ~group0_guard();
```

gleb-cloudius replied:
Even this is not needed IMO. Since topology_coordinator exists as a unique_ptr only I think the only thing that is needed is to make destructor available in the header.

piodul replied:
> WDYT about using pimpl pattern instead? Like with `group0_guard`.

I usually don't like the pimpl idiom because of the boilerplate that is needed for each exposed method, but considering that `topology_coordinator` only needs to expose `::run()` then it shouldn't be too bad.

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -114,11 +114,11 @@ class topology_coordinator_impl : public topology_coordinator {
                 // Remove from group 0 nodes that left. They may failed to do so by themselves
                 try {
-                    slogger.trace("raft topology: topology coordinator fiber removing {}"
+                    tlogger.trace("raft topology: topology coordinator fiber removing {}"
```
I think this should go along with removing the redundant `raft topology: ` prefix from all topology coordinator messages.

gleb-cloudius replied:
The non coordinator part also use the prefix and it is helpful to filter everything raft topology related.

gleb-cloudius commented:
I do not mind the change, but it will create a churn for everyone who have outstanding series. This definitely should wait for "add automatic sstable cleanup to the topology coordinator" series to be commuted.

kbr-scylla commented:
> This definitely should wait for "add automatic sstable cleanup to the topology coordinator" series to be commuted.

As you explained that series is blocked by host ID migration and we don't know for how long. So I don't think we should block this PR too. @piodul wants to continue implementing stuff on top (upgrade procedure)

gleb-cloudius commented:
> > This definitely should wait for "add automatic sstable cleanup to the topology coordinator" series to be commuted.
> 
> As you explained that series is blocked by host ID migration and we don't know for how long. So I don't think we should block this PR too. @piodul wants to continue implementing stuff on top (upgrade procedure)

This will cause a huge non mechanical rebase. I am firmly against this. The change here is mechanical and can be done at any moment later when the group effort will be in bug fixes and there will be not a lot of pending changes. @piodul can develop at the current placement and then mode the code as easy as now.

kbr-scylla commented:
The rebase is also mechanical. You don't apply the diff but take your modified topology coordinator code from storage_service and copy-paste it to replace the code in the new module. You have to do it for each commit but it's still mechanical.

avikivity commented:
> The rebase is also mechanical. You don't apply the diff but take your modified topology coordinator code from storage_service and copy-paste it to replace the code in the new module. You have to do it for each commit but it's still mechanical.

That's dangerous. Better to export the diff as a patch, sed the filenames, and apply it.

git sometimes guess file renames correctly.

piodul commented:
> The change here is mechanical and can be done at any moment later when the group effort will be in bug fixes and there will be not a lot of pending changes.

I actually had to do something like this today - rebase the code I wrote for this PR before I went to vacation. Recreating the last-but-one commit took 5-10 minutes and a simple copy-paste. It definitely won't be as easy for Gleb's and Petr's series, so I'm fine with waiting for a quieter moment.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5504/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6923/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5504/)
#### Failed Tests (27/23701):
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5504/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
#### Build Details:
- Duration: 1 hr 57 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5643/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7046/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5643/)
#### Failed Tests (27/23710):
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
- [test_topology_streaming_failure](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_topology_streaming_failure_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_streaming_failure)
- [test_tablet_drain_failure_during_decommission](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5643/testReport/junit/%28root%29/test_topology_failure_recovery/Tests___Unit_Tests___test_tablet_drain_failure_during_decommission_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_drain_failure_during_decommission)
#### Build Details:
- Duration: 2 hr 0 min
- Builder: spider3.cloudius-systems.com

gleb-cloudius commented:
It is a good time for v2.

piodul commented:
> It is a good time for v2.

I think I'll wait for https://github.com/scylladb/scylladb/pull/16749 to be merged as it conflicts with this PR.

piodul commented:
v2:
- Rebased
- Added the new files to `CmakeLists.txt`
- Changed the `topology_coordinator` to use the pimpl idiom

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5957/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7291/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5957/)
#### Build Details:
- Duration: 1 hr 46 min
- Builder: spider6.cloudius-systems.com

avikivity commented on service/topology_coordinator.hh:
```diff
@@ -0,0 +1,91 @@
+    void on_join_cluster(const gms::inet_address& endpoint) final {}
+    void on_leave_cluster(const gms::inet_address& endpoint) final {}
+    void on_up(const gms::inet_address& endpoint) final {}
+    void on_down(const gms::inet_address& endpoint) final;
```
Shouldn't these be virtual/override?

avikivity replied:
Pimlification + moving at the same time makes this hard to verify.

piodul replied:
> Shouldn't these be virtual/override?

I wanted to stick to the core guidelines: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-override

gleb-cloudius commented:
>     * Changed the `topology_coordinator` to use the pimpl idiom

What for though? It is instantiated into a `unique_ptr` by the storage service code anyway, so we have two inderections instead of one.

piodul commented:
> What for though?

My motivation was to reduce the need for recompilations. The topology coordinator is a large class which has a tiny public interface (basically, the `run` method and `endpoint_lifecycle_subscriber` interface). By having the `impl` hidden in the .cc file we can modify it more freely and not have to recompile storage_service every time.

I previously used a virtual interface but Kamil preferred a pimpl.

> It is instantiated into a `unique_ptr` by the storage service code anyway, so we have two inderections instead of one.

I removed the indirection through `unique_ptr` in storage service. The only `unique_ptr` now is inside the topology coordinator.

gleb-cloudius commented:
> > What for though?
> 
> My motivation was to reduce the need for recompilations. The topology coordinator is a large class which has a tiny public interface (basically, the `run` method and `endpoint_lifecycle_subscriber` interface). By having the `impl` hidden in the .cc file we can modify it more freely and not have to recompile storage_service every time.

What prevents you to do the same without pimpl. The only difference is that the storage service will be recompiled if internal topology coordinator interface changes and that does not happen often, but the implementation itself will be in a separate file.

> 
> I previously used a virtual interface but Kamil preferred a pimpl.
> 
> > It is instantiated into a `unique_ptr` by the storage service code anyway, so we have two inderections instead of one.
> 
> I removed the indirection through `unique_ptr` in storage service. The only `unique_ptr` now is inside the topology coordinator.

OK.

piodul commented:
> > > What for though?
> > 
> > 
> > My motivation was to reduce the need for recompilations. The topology coordinator is a large class which has a tiny public interface (basically, the `run` method and `endpoint_lifecycle_subscriber` interface). By having the `impl` hidden in the .cc file we can modify it more freely and not have to recompile storage_service every time.
> 
> What prevents you to do the same without pimpl. The only difference is that the storage service will be recompiled if internal topology coordinator interface changes and that does not happen often, but the implementation itself will be in a separate file.

I think it's debatable whether the internal interface changes often or not. The topology coordinator has quite a lot of private methods which were added over the course of many series/PRs. The overhead of pimpl or virtual interface is negligible in the topology coordinator's case, and IMO worth the time saved during the compilation.

gleb-cloudius commented:
> > > > What for though?
> > > 
> > > 
> > > My motivation was to reduce the need for recompilations. The topology coordinator is a large class which has a tiny public interface (basically, the `run` method and `endpoint_lifecycle_subscriber` interface). By having the `impl` hidden in the .cc file we can modify it more freely and not have to recompile storage_service every time.
> > 
> > 
> > What prevents you to do the same without pimpl. The only difference is that the storage service will be recompiled if internal topology coordinator interface changes and that does not happen often, but the implementation itself will be in a separate file.
> 
> I think it's debatable whether the internal interface changes often or not. The topology coordinator has quite a lot of private methods which were added over the course of many series/PRs. The overhead of pimpl or virtual interface is negligible in the topology coordinator's case, and IMO worth the time saved during the compilation.

That could be an argument to pimplify everything. Even if internal interface changes once in a while and storage service is recompiled it is not the end of the world. Let compile work instead of developers trying to figure out why an single implementation is hidden behind a redirection.

kbr-scylla commented:
> Let compile work instead of developers trying to figure out why an single implementation is hidden behind a redirection.

Topology coordinator already was behind a redirection (unique_ptr) as you observed. I like the shorter compile times. Let's not bikeshed over such minor decision.

gleb-cloudius commented:
> > Let compile work instead of developers trying to figure out why an single implementation is hidden behind a redirection.
> 
> Topology coordinator already was behind a redirection (unique_ptr) as you observed. I like the shorter compile times. Let's not bikeshed over such minor decision.

I was talking about code redirection obviously. The "shorter compile times" is bogus for 99.9% cases. Guys, not need to make thing more complicated than they should be.

kbr-scylla commented:
Code compilation is our issue #1 https://github.com/scylladb/scylladb/issues/1 and no it's not bogus, `storage_service.cc` is one of the longer compiling units. The fewer times we have to recompile it during development the better. There's nothing complicated about pimpl, it's a well known idiom.

gleb-cloudius commented:
> Code compilation is our issue number one #1 and no it's not bogus, `storage_service.cc` is one of the longer compiling units. The fewer times we have to recompile it during development the better. There's nothing complicated about pimpl, it's a well known idiom.

The issue is about compiling the code in general from scratch. Making this thing pimpl will not improve it even a little bit. On the contrary moving the code outside of  `storage_service.cc` (pimpl or not) will make it worse  since compiler will need to run per-processor twice now instead of once but still compile the same amount of code. (In the past there were even requests to put unrelated things into the same file because of it). So this is definitely not the argument you should make here, The argument is that changing topology coordinator privet API will not require recompilation of `storage_service.cc`  which is correct, but does it worth even a little bit of a code obfuscation? And the argument can be made for literally any C++ class with private functions. Do we start piplifying all of them now  to save some re-complications?

kbr-scylla commented:
> The argument is that changing topology coordinator privet API will not require recompilation of storage_service.cc which is correct, but does it worth even a little bit of a code obfuscation

Yes, please yes

> And the argument can be made for literally any C++ class with private functions. Do we start piplifying all of them now to save some re-complications?

Not sure, I think the usual counter-argument against pimpl is performance? If you're calling member functions on a given object often (on a hot path), each time requiring a pointer dereference, it might cause a visible degradation. On the other hand if you have a class like `topology_coordinator` with a method `run()` that is called once, this doesn't apply.

gleb-cloudius commented:
> > The argument is that changing topology coordinator privet API will not require recompilation of storage_service.cc which is correct, but does it worth even a little bit of a code obfuscation
> 
> Yes, please yesa

Are you compiling on a zx80? I did not see any speed issues with recompiling `storage_service.cc`

> 
> > And the argument can be made for literally any C++ class with private functions. Do we start piplifying all of them now to save some re-complications?
> 
> Not sure, I think the usual counter-argument against pimpl is performance? If you're calling member functions on a given object often (on a hot path), each time requiring a pointer dereference, it might cause a visible degradation. On the other hand if you have a class like `topology_coordinator` with a method `run()` that is called once, this doesn't apply.

There are a lot of classes that are not performance critical. Do we make them all pimpl? Anyway I see you have a strong opinion here. Let @avikivity  decide. If it wants more pimpl then be it. (but without pimpl you do not need to argue with him about virtual/override :))

kbr-scylla commented:
> There are a lot of classes that are not performance critical. Do we make them all pimpl?

Yes

gleb-cloudius commented:
> > There are a lot of classes that are not performance critical. Do we make them all pimpl?
> 
> Yes

Instead of fighting the language we chose to implement Scylla in may be just rewrite it one more time using something else? We will call it "getting rid of technical debt".

piodul commented:
@avikivity could you help us resolve the dispute about whether to use pimpl or not (starting here: https://github.com/scylladb/scylladb/pull/16609#issuecomment-1903473447)? I'd like to send an updated version only after I know which approach I should take.

gleb-cloudius commented:
> @avikivity could you help us resolve the dispute about whether to use pimpl or not (starting here: [#16609 (comment)](https://github.com/scylladb/scylladb/pull/16609#issuecomment-1903473447))? I'd like to send an updated version only after I know which approach I should take.

What about moving the `do_with` from `raft_state_monitor_fiber` to the new function "future<> run_topology_coordinator(...) inside your new file. Wouldn't it achieve the same encapsulation without pimpl?

piodul commented:
> What about moving the `do_with` from `raft_state_monitor_fiber` to the new function "future<> run_topology_coordinator(...) inside your new file. Wouldn't it achieve the same encapsulation without pimpl?

Yes, good idea. I suppose that, if it's later necessary to expose the topology coordinator, somebody else will do it and it will be their problem to choose how to do it best.

piodul commented:
v3:
- Rebased
- Hid the `topology_coordinator` in `topology_coordinator.cc`, now it can be only used via new function `run_topology_coordinator`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6011/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7340/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6011/)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: i-02fa65c0facd06156 (m5d.12xlarge)

piodul commented:
v3.1:
- Updated the cover letter
- Removed the definition of `raft_topology_cmd_handler_type` from storage_service.cc (it was defined in topology_coordinator.hh already)

nyh commented:
> > Code compilation is our issue number one #1 and no it's not bogus, `storage_service.cc` is one of the longer compiling units. The fewer times we have to recompile it during development the better. There's nothing complicated about pimpl, it's a well known idiom.
> 
> The issue is about compiling the code in general from scratch.

Not really - maybe this is what I wrote in the very first text in that issue, but it has been reimagined many times and my latest understanding of this issue is https://github.com/scylladb/scylladb/issues/1#issuecomment-1005191229 - as I explained there we have two things we need to optimize - **full compile** and **partial compile**. Some things we can do to improve both, but other things (like forward-definitions, pimpl, etc.) often helps partial compile more than full compile. But sometimes, you'll be surprised - even the full compile speed can be helped by reducing the use of complex templates and classes in header files and using pointers, forward declerations, and so on, instead.

> Making this thing pimpl will not improve it even a little bit. On the contrary moving the code outside of `storage_service.cc` (pimpl or not) will make it worse since compiler will need to run per-processor twice

The preprocessor probably takes %0.0001 (I didn't measure ;-)) of the compile time in our C++ code.

Anyway, the right way to do these things is to try and see what improves build speed and what doesn't. You can see some examples of tools used for this in the past in #1. You can measure build time directly, but in my experience, the size of generated object files ("*.o") is proportional to the build time, and it's often easier to measure.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6026/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7352/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6026/)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider6.cloudius-systems.com
